processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```
@@ -384,55 +381,115 @@ public int add(InputRow row) throws IndexSizeExceededException
-          dimensions.add(dimension);
+      }
+      if (dimensionValues.size() > 1) {
+        capabilities.setHasMultipleValues(true);
+      }
 
-          if (overflow == null) {
-            overflow = Lists.newArrayList();
-          }
-          overflow.add(getDimVals(dimValues.add(dimension), dimensionValues));
-        } else {
-          dims[index] = getDimVals(dimValues.get(dimension), dimensionValues);
+      Integer orderID = dimensionOrder.putIfAbsent(dimension, raceChecker);
+      while (orderID != null && orderID < 0) {
+        // Someone else is updating
+        orderID = dimensionOrder.get(dimension);
+      }
+      if (null == orderID) {
+        // We win race
+        orderID = orderIncrementer.getAndIncrement();
+        if (null == dims) {
+          dims = new String[orderIncrementer.get()][];
+        } else if (orderID >= dims.length) {
+          dims = Arrays.copyOf(dims, orderID + 1);
+        }
+        dims[orderID] = canonizeDimValues(dimValues.add(dimension), dimensionValues);
+        dimensionOrder.put(dimension, orderID);
+      } else {
+        final Integer index = getDimIndex(dimension);
+        if (null == dims) {
+          dims = new String[orderIncrementer.get()][];
+        } else if (index >= dims.length) {
+          dims = Arrays.copyOf(dims, orderIncrementer.get());
         }
+        DimDim values;
+        do {
+          // Race condition on concurrent writes
+          values = dimValues.get(dimension);
+        } while (null == values);
+        dims[index] = canonizeDimValues(values, dimensionValues);
       }
     }
 
-    if (overflow != null) {
-      // Merge overflow and non-overflow
-      String[][] newDims = new String[dims.length + overflow.size()][];
-      System.arraycopy(dims, 0, newDims, 0, dims.length);
-      for (int i = 0; i < overflow.size(); ++i) {
-        newDims[dims.length + i] = overflow.get(i);
-      }
-      dims = newDims;
+    if (null == dims) {
```

how can this happen? should we throw an error here instead?